### ISSUE TITLE
Improvement proposal: Don't rely on speedtest-cli for speed measurements

### ISSUE THREAD

**Issue by momu** [ansemjo/speedtest-plotter#4]
Thanks for the easy to use tool. I run it on my NAS but figured out that it is not very reliable when it comes to download speeds.

Below is my plot of the last days. 

![Screenshot_20200404_113026](https://user-images.githubusercontent.com/1451642/78423614-bfcfd800-7667-11ea-939c-49c3c0d8fcbb.png)

I have a 50/10 Mbit connection. Especially when the download speed was slow I double checked with other browser based services like 

* http://avm.de/nc/service/zack-der-speedtest-fuer-ihre-breitbandverbindung/
* https://speedof.me/
* https://www.speedtest.net/

Even speedtest.net showed higher results when using the browser based service.

Furthermore I did some tests with 

```
wget -O /dev/null http://speedtest.tele2.net/200MB.zip
```
Again the readings were higher.

Checking the `speedtest-cli` [README](https://github.com/sivel/speedtest-cli#inconsistency) 
gives:

> There is the potential for this tool to report results inconsistent with Speedtest.net. There are several concepts to be aware of that factor into the potential inconsistency:
> 
>1. Speedtest.net has migrated to using pure socket tests instead of HTTP based tests
>2. This application is written in Python
>3. Different versions of Python will execute certain parts of the code faster than others
>4. CPU and Memory capacity and speed will play a large part in inconsistency between Speedtest.net and even other machines on the same network
> 
> Issues relating to inconsistencies will be closed as wontfix and without additional reason or context.

The http based test in python seems to be not very reliable. 

Some further details here:

https://superuser.com/questions/1434610/speedtest-cli-very-slow-although-network-speed-is-fine/1519159#1519159

I suggest to change the tool for speed measurement to either one of them:

* https://www.speedtest.net/apps/cli
* SpeedTest++ https://github.com/taganaka/SpeedTest
* `wget` or `curl` based test with e.g. http://speedtest.tele2.net/

Another interesting fact: 

Comparing `ftp` based download speed (`wget -O /dev/null ftp://speedtest.tele2.net/100MB.zip --report-speed=bits`) with `http` based speed (`wget -O /dev/null http://speedtest.tele2.net/100MB.zip --report-speed=bits`) shows that `http` is about 3x faster (`45.9Mb/s` vs `17.5Mb/s`).

There were rumors some years ago that some providers prioritize anything which has `speedtest` in it - but only for `http`. 
So it could make sense to include `ftp` speedtest as well.

**Comment by ansemjo** [comment]
Thanks for your input, that is interesting information ..

I did observe that tests against speedtest.net are consistently faster when run in the browser, too. But I always kind of attributed it to other factors like picking the server purely on the fastest ping or bottlenecks in the way I ran the container. Back when I had a 60 Mbps downlink from Vodafone I never had these discrepancies.

As I mentioned recently in https://github.com/ansemjo/speedtest-plotter/issues/2#issuecomment-604494755 this tool was not intended to be what it is today in the beginning. The choice of running `speedtest-cli` in a subprocess instead of using the Python bindings and then generating the plot by piping in the PNG output from `gnuplot` .. well. :P Improving on these inconsistencies you mentioned might be a good reason to finally do a "v2" with client-side graph rendering, too.

---

Of the choices you mentioned, I would not use the purely wget-based downloads from a specific server .. for one thing, there probably is a reason why speedtest.net moved away from HTTP based testing. On the other hand, choosing a specific server is going to lead to very different results around the globe. Speedtest.net's network is hard to beat in this regard. (FYI: the LeaseWeb mirror has some pretty large speedtest files of up to 10 GB, too: https://mirror.leaseweb.net/)

I'm not sure about Ookla's official client either .. for one thing I'm not sure how I can install it on Alpine Linux for the Docker container. And I am not sure about the licensing, if I am allowed to do that at all.

I'll take a look at speedtest++ though. Maybe it's a drop-in replacement for speedtest-cli ..

**Comment by ansemjo** [comment]
So I checked SpeedTest++ and yes, it does give drastically higher download speeds. It's not completely drop-in because it has a different output format. But it should be trivial to parse its JSON output.

I also found [ddsol/speedtest.net](https://github.com/ddsol/speedtest.net), which "circumvents" the EULA restriction in a clever way: it downloads the official binary dynamically and requires you to first accepting the terms with special commandline flags.

**Comment by ansemjo** [comment]
I've created a [new branch](https://github.com/ansemjo/speedtest-plotter/tree/altclient) and amended the Dockerfile to build the necessary SpeedTest binary in a multistage build, which I've pushed manually to [ansemjo/speedtest:altclient](https://hub.docker.com/r/ansemjo/speedtest/tags) just now.

The database format hasn't changed, so depending on how you persisted your database, you should be able to pull this new tag and just check it out:

```
docker pull ansemjo/speedtest:altclient
docker rm -f $oldcontainer
docker run -d ... -v /path/to/data:/data ansemjo/speedtest:altclient
```

I am getting **much** better results with this new client. Depending on how your results turn out I might just tidy it up a bit and merge this change into master.

**Comment by momu** [comment]
I just pulled the new tag and the first two results are looking better. Will get back to you with some results tomorrow.

**Comment by momu** [comment]
I also see that I get much better results with `SpeedTest++`, below are the results. I let run `speedtest-cli` one minute before `SpeedTest++`. So they should give the same results. Again `SpeedTest++` are similar to the some browser based tests. Below each graph is a picture of the up- and downstream in my router. `SpeedTest++` takes around `60 s`, but I also observed durations of `40 s`.

Long story short: speed test results are much better now. Thanks for the fast implementation.

I thought about changing it myself but figured out very soon that it would end in a never ending story.

## speedtest-cli

![image](https://user-images.githubusercontent.com/1451642/78644376-129ecf00-78b6-11ea-90db-9fe5dd9122b1.png)

![Screenshot_20200407_100150](https://user-images.githubusercontent.com/1451642/78644965-0b2bf580-78b7-11ea-9a27-d1d66249260b.png)

## SpeedTest++

![image](https://user-images.githubusercontent.com/1451642/78644412-1fbbbe00-78b6-11ea-8cb1-9445073114e8.png)

![image](https://user-images.githubusercontent.com/1451642/78644942-ff403380-78b6-11ea-89ab-e2c24de8641a.png)

**Comment by moeffju** [comment]
As an additional data point, I just spun up the altclient in parallel to the original image, using a different volume but on the same docker host, and it’s a night and day difference. Original measured ~250 Mbit/s down, altclient gets 930 Mbit/s.

**Comment by ansemjo** [comment]
Yes, I understand. I have a Vodafone Gigabit connection at home myself and I had a difference of ~430 Mbps to ~950 Mbps, too.

**Comment by ansemjo** [comment]
FYI: the alternate client is now merged into the main branch (5511024). So `ansemjo/speedtest` is now using SpeedTest++.
